File: Top_Interview_150/Array_String/cpp/68.cpp
#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    vector<string> fullJustify(vector<string> &words, int maxWidth)
    {
        vector<pair<vector<string>, int>> strings = this->greedy_fill(words, maxWidth);

        vector<string> response;
        for (int i = 0; i < strings.size() - 1; ++i)
        {
            vector<string> data = strings[i].first;
            int length = data.size();
            int count = strings[i].second;
            string cur_string = data[0];

            int space_places = length - 1;
            if (space_places == 0)
            {
                cur_string.append(count, ' ');
                response.push_back(cur_string);
                continue;
            }

            int count_spaces = count / space_places;
            int ost_spaces = count - (count_spaces * space_places);

            for (int i = 1; i < length; ++i)
            {
                cur_string.append(count_spaces, ' ');
                if (ost_spaces)
                {
                    cur_string += ' ';
                    ost_spaces--;
                }
                cur_string += data[i];
            }

            response.push_back(cur_string);
        }

        vector<string> data = strings[strings.size() - 1].first;
        int length = data.size();
        int count = strings[strings.size() - 1].second;

        string cur_string;
        for (int i = 0; i < length - 1; ++i)
        {
            cur_string += data[i] + ' ';
        }

        cur_string += data[length - 1];
        cur_string.append(count - length + 1, ' ');
        response.push_back(cur_string);

        return response;
    }

    vector<pair<vector<string>, int>> greedy_fill(vector<string> &words, int maxWidth)
    {
        int n = words.size();

        vector<pair<vector<string>, int>> result;
        vector<string> cur_data = {words[0]};
        int cur_size = words[0].size();
        for (int i = 1; i < n; ++i)
        {
            string word = words[i];
            if (cur_size + 1 + word.size() <= maxWidth)
            {
                cur_data.push_back(word);
                cur_size += 1 + word.size();
            }
            else
            {
                result.push_back({cur_data, maxWidth - (cur_size - cur_data.size() + 1)});
                cur_data = {word};
                cur_size = word.size();
            }
        }

        if (cur_size)
            result.push_back({cur_data, maxWidth - (cur_size - cur_data.size() + 1)});

        return result;
    }
};

File: Top_Interview_150/Array_String/python/68.py
from typing import List, Tuple


class Solution:

    @staticmethod
    def greedy_fill(words: List[str], maxWidth: int) -> List[Tuple]:
        n = len(words)

        result = []
        cur_data = [words[0]]
        cur_size = len(words[0])
        for i in range(1, n):
            word = words[i]
            if cur_size + 1 + len(word) <= maxWidth:
                cur_data.append(word)
                cur_size += 1 + len(word)
            else:
                result.append(
                    (cur_data, maxWidth - (cur_size - len(cur_data) + 1)))
                cur_data = [word]
                cur_size = len(word)

        if cur_size:
            result.append(
                (cur_data, maxWidth - (cur_size - len(cur_data) + 1)))

        return result

    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:

        strings = self.greedy_fill(words, maxWidth)

        response = []
        for i in range(len(strings) - 1):
            data, count = strings[i]
            cur_string = data[0]

            space_places = len(data) - 1
            if space_places == 0:
                cur_string += count * " "
                response.append(cur_string)
                continue

            count_spaces = count // space_places
            ost_spaces = count - (count_spaces * space_places)

            for i in range(1, len(data)):
                cur_string += " " * count_spaces
                if ost_spaces:
                    cur_string += " "
                    ost_spaces -= 1
                cur_string += data[i]

            response.append(cur_string)

        data, count = strings[-1]
        response.append(" ".join(data) + " " * (count - len(data) + 1))

        return response


solution = Solution()
print(solution.fullJustify(words=["Science", "is", "what", "we", "understand", "well", "enough", "to",
      "explain", "to", "a", "computer.", "Art", "is", "everything", "else", "we", "do"], maxWidth=20))
